### PR TITLE
hatch 1.6.2

### PR DESCRIPTION
**Pull request by Porkepix**
Created by `brew bump`

---

Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by Porkepix** [comment]
I'm really confused, error is the same as in https://github.com/Homebrew/homebrew-core/pull/112707 but was not there in https://github.com/Homebrew/homebrew-core/pull/113375

**Comment by michaelherold** [comment]
What platform did you run the command on?

For https://github.com/Homebrew/homebrew-core/pull/113375, I ran (on a 2019 Macbook Pro, macOS 12.3.1):

    brew bump-formula-pr --write-only --url=https://files.pythonhosted.org/packages/ff/61/92f38381f1cd234faf76da71d67b506ccd215d1cad5d476450b745739d90/hatch-1.6.1.tar.gz --sha256 2e50fd6edcc1de5a4c7e9b3b820a1bf61505a09bedab642f081dcc1acbbf6c44 hatch

because of issues with `brew bump`. I followed up with:

    brew install --build-from-source hatch
    brew audit hatch
    brew test hatch
    brew audit --strict --online hatch

I also manually reviewed the dependencies that were bumped because `brew bump` was unhappy. I'm a neophyte Homebrew contributor, but I'm not new to packaging in general so I have a general distrust of automated bumping. 😬

I'm currently running an install bullt from source with this patch minus the `cffi`, `cryptography`, `jeepney`, `pycparser`, and `SecretStorage` dependencies to see how that works.

**Comment by Porkepix** [comment]
Command was run from a Linux computer.

Many formulas including those with many Python resources works well with the automatic bump, so I guess there's a way for that one to don't encounter such issue too?

**Comment by michaelherold** [comment]
> I'm currently running an install bullt from source with this patch minus the cffi, cryptography, jeepney, pycparser, and SecretStorage dependencies to see how that works.

The install works correctly like this on macOS. With [the example](https://github.com/Homebrew/homebrew-core/blob/1a727bed0524f15e0a8ac8a992d03ac39f6e7c47/Formula/organize-tool.rb#L94-L100) linked in [this comment](https://github.com/Homebrew/homebrew-core/pull/112707#issuecomment-1275525431) on the bump for 1.6.0, I'm wondering if it would make sense to have something like this in the package:

```ruby
def install
  venv = virtualenv_create(libexec, "python3.10")
  dependencies = resources.map(&:name).to_set
  unless OS.linux?
    # `SecretStorage` (and `jeepney`) and its dependencies are only needed on Linux
    dependencies -= %w[cffi cryptography jeepney pycparser SecretStorage]
  end
  dependencies.each do |r|
    venv.pip_install resource(r)
  end
  venv.pip_install_and_link buildpath
end
```

The downside here is that since `SecretStorage` is a transitive dependency, we have a moving target that isn't directly observable when looking at Hatch release notes.

Also, I'm unclear what happens if you bump on macOS; would these dependencies be bumped or left alone on that platform? My intuition says they would be left alone, but I'm unsure whether that is true or not.

**Comment by michaelherold** [comment]
I dug into this a little more and think I may have found a solution. For Linux-only configuration, you can use `on_linux` to specify that macOS does not handle them. When you place the new resources in an `on_linux` block, they _should_ not apply to macOS.

See the patch below:

<details>
<summary>Patch to move Linux-only dependencies into an <code class="notranslate">on_linux</code> block</summary>

```diff
From da66f08f0bd6f2fcd8df48e9575b3a090b752021 Mon Sep 17 00:00:00 2001
From: Michael Herold <opensource@michaeljherold.com>
Date: Sun, 23 Oct 2022 22:09:49 -0500
Subject: [PATCH] Adjust Linux-only dependencies

Some transitive dependencies [through keyring][1] are only needed on
Linux. As such, they should be marked that way to ensure macOS does not
try to install them.

[1]: https://github.com/jaraco/keyring/blob/f9142e2ab8f237708a88c52d0fc4cf4f1cba50b2/setup.cfg#L24-L25
---
 Formula/hatch.rb | 52 +++++++++++++++++++++++++-----------------------
 1 file changed, 27 insertions(+), 25 deletions(-)

diff --git a/Formula/hatch.rb b/Formula/hatch.rb
index 0d44631006d..e5e95a02bf8 100644
--- a/Formula/hatch.rb
+++ b/Formula/hatch.rb
@@ -30,11 +30,6 @@ class Hatch < Formula
     sha256 "0d9c601124e5a6ba9712dbc60d9c53c21e34f5f641fe83002317394311bdce14"
   end
 
-  resource "cffi" do
-    url "https://files.pythonhosted.org/packages/2b/a8/050ab4f0c3d4c1b8aaa805f70e26e84d0e27004907c5b8ecc1d31815f92a/cffi-1.15.1.tar.gz"
-    sha256 "d400bfb9a37b1351253cb402671cea7e89bdecc294e8016a707f6d1d8ac934f9"
-  end
-
   resource "click" do
     url "https://files.pythonhosted.org/packages/59/87/84326af34517fca8c58418d148f2403df25303e02736832403587318e9e8/click-8.1.3.tar.gz"
     sha256 "7682dc8afb30297001674575ea00d1814d808d6a36af415a82bd481d37ba7b8e"
@@ -45,11 +40,6 @@ class Hatch < Formula
     sha256 "452f9dc859be7f06631ddcb328b6919c67984aca654e5fefb3914d54691aed60"
   end
 
-  resource "cryptography" do
-    url "https://files.pythonhosted.org/packages/63/82/a6e21842f2e31b3874f01c112093b8bf8af119f5ed999bbd667a81de720b/cryptography-38.0.2.tar.gz"
-    sha256 "7a022ec87c7a8bdad99f516a4ee6ffcb3a2bc31487577f9eccbc9b2edb1a8fd4"
-  end
-
   resource "distlib" do
     url "https://files.pythonhosted.org/packages/58/07/815476ae605bcc5f95c87a62b95e74a1bce0878bc7a3119bc2bf4178f175/distlib-0.3.6.tar.gz"
     sha256 "14bad2d9b04d3a36127ac97f30b12a19268f211063d8f8ee4f47108896e11b46"
@@ -100,11 +90,6 @@ class Hatch < Formula
     sha256 "89559fa5c1d3c34eff6f631ad80bb21f378dbcbb35dd161fd2c6b93f5be2f98a"
   end
 
-  resource "jeepney" do
-    url "https://files.pythonhosted.org/packages/d6/f4/154cf374c2daf2020e05c3c6a03c91348d59b23c5366e968feb198306fdf/jeepney-0.8.0.tar.gz"
-    sha256 "5efe48d255973902f6badc3ce55e2aa6c5c3b3bc642059ef3a91247bcfcc5806"
-  end
-
   resource "keyring" do
     url "https://files.pythonhosted.org/packages/2a/ef/28d3d5428108111dae4304a2ebec80d113aea9e78c939e25255425d486ff/keyring-23.9.3.tar.gz"
     sha256 "69b01dd83c42f590250fe7a1f503fc229b14de83857314b1933a3ddbf595c4a5"
@@ -145,11 +130,6 @@ class Hatch < Formula
     sha256 "5c5d0a3b48ceee0b48485e0c26037c0acd7d29765ca3fbb5cb3831d347423220"
   end
 
-  resource "pycparser" do
-    url "https://files.pythonhosted.org/packages/5e/0b/95d387f5f4433cb0f53ff7ad859bd2c6051051cebbb564f139a999ab46de/pycparser-2.21.tar.gz"
-    sha256 "e644fdec12f7872f86c58ff790da456218b10f863970249516d60a5eaca77206"
-  end
-
   resource "pyparsing" do
     url "https://files.pythonhosted.org/packages/71/22/207523d16464c40a0310d2d4d8926daffa00ac1f5b1576170a32db749636/pyparsing-3.0.9.tar.gz"
     sha256 "2b020ecf7d21b687f219b71ecad3631f644a47f01403fa1d1036b0c6416d70fb"
@@ -170,11 +150,6 @@ class Hatch < Formula
     sha256 "ba3a3775974105c221d31141f2c116f4fd65c5ceb0698657a11e9f295ec93fd0"
   end
 
-  resource "SecretStorage" do
-    url "https://files.pythonhosted.org/packages/53/a4/f48c9d79cb507ed1373477dbceaba7401fd8a23af63b837fa61f1dcd3691/SecretStorage-3.3.3.tar.gz"
-    sha256 "2403533ef369eca6d2ba81718576c5e0f564d5cca1b58f73a8b23e7d4eeebd77"
-  end
-
   resource "shellingham" do
     url "https://files.pythonhosted.org/packages/bd/e6/fdf53ebbf08016dba98f2b047d4db95790157f0e2eed3b14bb5754271475/shellingham-1.5.0.tar.gz"
     sha256 "72fb7f5c63103ca2cb91b23dee0c71fe8ad6fbfd46418ef17dbe40db51592dad"
@@ -210,6 +185,33 @@ class Hatch < Formula
     sha256 "227ea1b9994fdc5ea31977ba3383ef296d7472ea85be9d6732e42a91c04e80da"
   end
 
+  on_linux do
+    resource "cffi" do
+      url "https://files.pythonhosted.org/packages/2b/a8/050ab4f0c3d4c1b8aaa805f70e26e84d0e27004907c5b8ecc1d31815f92a/cffi-1.15.1.tar.gz"
+      sha256 "d400bfb9a37b1351253cb402671cea7e89bdecc294e8016a707f6d1d8ac934f9"
+    end
+
+    resource "cryptography" do
+      url "https://files.pythonhosted.org/packages/63/82/a6e21842f2e31b3874f01c112093b8bf8af119f5ed999bbd667a81de720b/cryptography-38.0.2.tar.gz"
+      sha256 "7a022ec87c7a8bdad99f516a4ee6ffcb3a2bc31487577f9eccbc9b2edb1a8fd4"
+    end
+
+    resource "jeepney" do
+      url "https://files.pythonhosted.org/packages/d6/f4/154cf374c2daf2020e05c3c6a03c91348d59b23c5366e968feb198306fdf/jeepney-0.8.0.tar.gz"
+      sha256 "5efe48d255973902f6badc3ce55e2aa6c5c3b3bc642059ef3a91247bcfcc5806"
+    end
+
+    resource "pycparser" do
+      url "https://files.pythonhosted.org/packages/5e/0b/95d387f5f4433cb0f53ff7ad859bd2c6051051cebbb564f139a999ab46de/pycparser-2.21.tar.gz"
+      sha256 "e644fdec12f7872f86c58ff790da456218b10f863970249516d60a5eaca77206"
+    end
+
+    resource "SecretStorage" do
+      url "https://files.pythonhosted.org/packages/53/a4/f48c9d79cb507ed1373477dbceaba7401fd8a23af63b837fa61f1dcd3691/SecretStorage-3.3.3.tar.gz"
+      sha256 "2403533ef369eca6d2ba81718576c5e0f564d5cca1b58f73a8b23e7d4eeebd77"
+    end
+  end
+
   def install
     virtualenv_install_with_resources
 
-- 
2.32.1 (Apple Git-133)
```
</details>

@Porkepix, could you apply that patch and see if it helps?

**Comment by Porkepix** [comment]
> The downside here is that since `SecretStorage` is a transitive dependency, we have a moving target that isn't directly observable when looking at Hatch release notes.

I think I don't really understand the meaning of that sentence, but it's my English for that I'm not a native speaker to blame for.

We'll see with your patch I just applied, however I'm not sure how will it behave in the future in case of other dependencies changing and causing issues on one platform or the other.

**Comment by michaelherold** [comment]
> I think I don't really understand the meaning of that sentence, but it's my English for that I'm not a native speaker to blame for.

That's fair! It's a needlessly complex sentence, my apologies. Let me try again.

With this fix, I'm not sure what happens when you bump the library from either platform. When bumping from macOS, I worry that the bump will not re-evaluate the Linux-only dependencies. When bumping from Linux, I worry that new dependencies that are Linux-only will creep into the resource list outside the `on_linux` block.

I suppose the latter will be caught by CI, so it's less of a worry just an annoyance. But I am concerned about the former as it will be a maintenance burden.

I guess it's up to the Homebrew maintainers to weigh in!

**Comment by Porkepix** [comment]
> > I think I don't really understand the meaning of that sentence, but it's my English for that I'm not a native speaker to blame for.
> 
> That's fair! It's a needlessly complex sentence, my apologies. Let me try again.
> 
> With this fix, I'm not sure what happens when you bump the library from either platform. When bumping from macOS, I worry that the bump will not re-evaluate the Linux-only dependencies. When bumping from Linux, I worry that new dependencies that are Linux-only will creep into the resource list outside the `on_linux` block.
> 
> I suppose the latter will be caught by CI, so it's less of a worry just an annoyance. But I am concerned about the former as it will be a maintenance burden.
> 
> I guess it's up to the Homebrew maintainers to weigh in!

Thanks for the explanation! Not too worried for the bump from Linux, what I'm more worried about is that doing so from Mac would maybe completely remove these resources as `pipgrip` won't list them.

**Review by SMillerDev — APPROVED** [review]
